apollo-router/src/drop_watch.rs

```diff
@@ -0,0 +1,141 @@
+//! Provide a [`DropWatch`] utility.
+//!
+//! Accept a FnOnce and carefully watch over it until we are dropped.
+//!
+//! Useful when we want to join threads which will ultimately be terminated
+//! within an asynchronous context.
+
+use std::mem::ManuallyDrop;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::thread::JoinHandle;
+
+pub(crate) struct DropWatch {
+    watcher_handle: ManuallyDrop<JoinHandle<()>>,
+    park_flag: Arc<AtomicBool>,
+}
+
+impl DropWatch {
+    /// Create a DropWatch which will execute the supplied closure then wait for notification
+    pub(crate) fn run_and_wait<F: FnOnce() + Send + 'static>(watched: F) -> Self {
+        let park_flag = Arc::new(AtomicBool::new(false));
```

Comment:
How about using a tokio mutex? Create an owned lock guard and pass it into the spawned thread.
Then try to lock it in the original thread. This would remove the thread park stuff.

Reply:
Because you can't `join` it. You have to `await`. This is specifically for code which is not async and which we would like to know it completed.